features/370-auth-ui/design.md

```diff
@@ -0,0 +1,393 @@
+   * [The signup password page](#the-signup-password-page)
+* [The settings endpoint](#the-settings-endpoint)
+* [The verification endpoint](#the-verification-endpoint)
+* [The logout endpoint](#the-logout-endpoint)
+* [Omitted features](#omitted-features)
+* [Future improvements](#future-improvements)
+   * [Customizing the signup flow or the authentication flow](#customizing-the-signup-flow-or-the-authentication-flow)
+* [Support Matrix](#support-matrix)
+* [Being a OpenID Provider](#being-a-openid-provider)
+   * [Authorization Endpoint](#authorization-endpoint)
+   * [Token Endpoint](#token-endpoint)
+   * [Token Response](#token-response)
+* [Security considerations](#security-considerations)
+   * [Being embedded in an HTML iframe](#being-embedded-in-an-html-iframe)
+* [Appendix](#appendix)
+
+## Background
+
+Auth UI is the builtin web frontend of Auth Gear.
+It provides access to most Auth Gear features as a server side rendered web application.
+It is part of Auth Gear.
+
+It is intended to be a conforming OpenID Provider implementation of Auth Gear in the future.
+As of the time of writing, the conformance of OpenID Provider is not a goal.
+However, the implementation must be prepared for conformance in the future.
+
+## Where Auth UI is served
+
+It is served in the same domain as Auth Gear and Microservices.
+
+### Reason 1: Individual gears should be unaware of Custom Domain
+
+Auth UI is going to be responsible for user verification. The verification email includes an URL to Auth UI. This implies Auth Gear has to know where Auth UI is served. When Custom Domain was introduced, it is a feature solely handled by Gateway. Individual gears such as Auth Gear are unaware of Custom Domain. Serving Auth UI in a different domain will break this property. If Auth UI is served in the same domain, it is trivial for Auth Gear to know where it is.
+
+### Reason 2: Cookies written by Auth Gear is host-only
+
+A host-only cookie only matches the domain by simple string comparison. While a non host-only cookie matches the domain via the [domain-matches](https://tools.ietf.org/html/rfc6265#section-5.1.3) algorithm. To allow Auth UI to be embedded within an iframe in a Skygear App, the session and its cookie must be shared by both. This means the domain has to be either exact match or domain-matches.
+
+If cookies written by Auth Gear have to be not host-only, then the cookie attribute `Domain` must be specified. If the custom domain is `www.example.com`, then the value of the attribute should be the apex domain `example.com`. So Auth Gear cannot avoid being aware of Custom Domain.
+
+### The actual location of Auth UI
+
+Auth UI is served at the path prefix `/_auth/_/`
+
+Some example endpoints:
+
+- `/_auth/_/authorize`
+- `/_auth/_/signup`
+- `/_auth/_/forgot_password`
+- `/_auth/_/reset_password`
+- `/_auth/_/verify`
+
+As a side note, Universal Login of Auth0 is served at `https://<app>.auth0.com/u/{login,signup,...}`
```

Comment:
Auth0 support custom domain too (https://auth0.com/docs/custom-domains)

Anyway... but I feel like it is werid to include it here..

Reply:
Let's continue the discussion in https://github.com/SkygearIO/features/pull/411#discussion_r378191375